### PR TITLE
App Lab: Fix touch

### PR DESCRIPTION
**Pull request by breville**
This is a followup to https://github.com/code-dot-org/code-dot-org/pull/34164 to get it working on iOS devices.  From some very rudimentary print-to-screen debugging, it appeared that [modifying a read-only property of a Touch event](https://github.com/code-dot-org/code-dot-org/pull/34164/files#diff-62379d80b2be88365b79e987c324d831R117) was triggering an exception on iOS devices and no further code was executed.

In this fix, I refactor the code a bit to treat the event as immutable.

To test, I built an App Lab app which supports both drawing (using a mouse or touch) and a button click:

```
hide();
penColor("green");
penWidth(5);
penUp();

onEvent("button1", "click", function(event) {
  penColor("red");
  textLabel("label1", "clicked");
});

onEvent("screen1", "mousemove", function(event) {
  moveTo(event.x, event.y);
});
onEvent("screen1", "mousedown", function(event) {
  moveTo(event.x, event.y);
  penDown();
});
onEvent("screen1", "mouseup", function(event) {
  penUp();
});
```

It appears to operate successfully both in the App Lab editor and as a shared page, on both desktop and iOS device.

I did come across a couple other existing issues that are not addressed here:

- If the share view is scrolled down, the ink is drawn above the cursor/touch point because we don't appear to compensate for the page's scroll offset.
- While drawing on an iOS device, it also causes the page to scroll a bit.  I attempted to call `preventDefault` for [touch events](https://github.com/code-dot-org/code-dot-org/pull/34215/files#diff-62379d80b2be88365b79e987c324d831R112-R114), and while it prevented scrolling, it also prevented the app's button's click event from firing.  I also looked into adding an `overflow: hidden` to the page's [`<html>`](https://github.com/code-dot-org/code-dot-org/blob/534e0d378d70b2f8e3ab12444983a64066e0c1e2/dashboard/app/views/layouts/application.html.haml#L2) element, and while it also prevented any page movement from occuring, I don't think we can make an assumption that the user doesn't want to scroll since the share view doesn't appear to shrink the page's content to fit the user's vertical screen space, at least on my desktop browser.

### CODE REVIEW

**Comment by jmkulwik** [comment]
Tested on Android and Surface, and both worked great ~(Note for anyone else testing: You have to add `hide()` to the beginning of the program. Button clicking doesn't work with the turtle visible)~ Fixed in the description!

File: apps/src/applab/EventSandboxer.js [hunk 1]
```diff
@@ -99,63 +99,68 @@ EventSandboxer.prototype.sandboxEvent = function(event) {
...
-    if (typeof event[prop] !== 'undefined') {
-      newEvent[prop] = (event[prop] - this.yOffset_) / this.yScale_;
+    if (typeof mouseEvent[prop] !== 'undefined') {
+      newEvent[prop] = (mouseEvent[prop] - this.yOffset_) / this.yScale_;
     }
   }, this);
+
+  // Add values that can be missing when touch is used.
+  if (mouseEvent.x === undefined) {
+    newEvent.x = (mouseEvent.pageX - this.xOffset_) / this.xScale_;
```
Comment (jmkulwik) [inline apps/src/applab/EventSandboxer.js]: Are we guaranteed to have a `pageX` in all cases? i.e. keyUp/keyDown? Suggestion: when you create `mouseEvent` from `event.changedTouches[0]` above, assign the x and y. i.e.  `mouseEvent.x = event.changedTouches[0].pageX`

**Comment by jmkulwik** [comment]
Looks great! I have one quick comment (see code). Besides that, awesome work and cleanup!

**Review by jmkulwik — APPROVED** [review]

File: apps/src/applab/EventSandboxer.js (hunk 1 above)
Comment (breville) [reply apps/src/applab/EventSandboxer.js]: That's a good point, and I will rearrange the code . That said, once assigned a value, `mouseEvent` will be a [`TouchEvent`](https://developer.mozilla.org/en-US/docs/Web/API/Touch) or some other kind of [`Event`](https://developer.mozilla.org/en-US/docs/Web/API/Event) and I want to treat it as an immutable, read-only object as well.  I want to write only into `newEvent` which is the local object that is designed to receive all the calculated values.

In fact, it appears that MouseEvent.x is just an alias for MouseEvent.clientX, so we could modify the other App Lab code that depends on this value, but I want to keep this change localised.

**Pull request merged**